pennylane/decomposition/__init__.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+r"""
+.. currentmodule:: pennylane
+
+This module implements the infrastructure for PennyLane's new graph-based decomposition system.
+
+.. warning::
+
+    This module is experimental and is subject to change in the future.
+
+.. currentmodule:: pennylane.decomposition
+
+.. autosummary::
+    :toctree: api
+
+    ~register_resources
+    ~resource_rep
+    ~controlled_resource_rep
+    ~adjoint_resource_rep
+    ~add_decomps
+    ~list_decomps
+    ~has_decomp
+    ~Resources
+    ~CompressedResourceOp
+    ~DecompositionRule
+
+.. TODO::
+    Add section here explaining how to enable and disable the new decomposition system. [sc-83993]
+
+**Defining Decomposition Rules**
+
+In the new decomposition system, a decomposition rule must be defined as a quantum function that
+accepts ``(*op.parameters, op.wires, **op.hyperparameters)`` as arguments, where ``op`` is an
+instance of the operator type that the decomposition is for. Additionally, a resource estimate
+in the form of a gate count must be registered with this function:
+
+.. code-block:: python
+
+    import pennylane as qml
+
+    @qml.register_resources({qml.H: 2, qml.CZ: 1})
+    def my_cnot(wires):
+        qml.H(wires=wires[1])
+        qml.CZ(wires=wires)
+        qml.H(wires=wires[1])
+
+.. code-block:: pycon
+
+    >>> with qml.queuing.AnnotatedQueue() as q:
+    ...     my_cnot(wires=[0, 1])
+    >>> q.queue
+    [H(1), CZ(wires=[0, 1]), H(1)]
+    >>> my_cnot.compute_resources()
+    num_gates=3, gate_counts={H: 2, CZ: 1}
```

Comment:
I suggest putting a more user-friendly example here, or nothing at all (and removing the above `register_resources` example).